### DISCUSSION TITLE
Save trice logs on target

### DISCUSSION THREAD

**Discussion by FabienLep** [rokath/trice#405] [Q&A] [answered]
Hi,
Thank you for the great project! I really like how trice is built and would like to use it on my project.
I am looking into saving logs directly on target (in a flash file system or RAM circular buffer) to be able to upload the logs to a server via LTE or BLE and analyse them offline.
I looked through the code and found some "hacks" to save the logs on device instead of direct UART/RTT output but would like your input on the proper way to do this?
https://github.com/rokath/trice/discussions/333 looked promising but I don't think TRICE_WRITE is supported anymore?
Do you have input on the proper/cleanest way to do this?
Thanks!

**Comment by rokath** [comment, marked as answer]
Sorry, Fabien, for the delay. I only saw your question right now.

In the development branch look for `TRICE_AUXILIARY`. This is not the final state but shows the general idea. I have no universal solution for this right now. The point is: For each output channel the user should be able to choose between direct and deferred mode. I am trying to implement that in the experimental branch.

For your purpose you could extend `TriceWriteDevice` in [core.c](https://github.com/rokath/trice/blob/master/src/box/core.c) in the master branch.

In the future I will provide a function pointer variable for this.

**Reply by FabienLep** [reply]
Thanks! Things look good on development branch to start from there.

**Comment by rokath** [comment]
Branch *development* right now merged into master (and released as [v0.57.0](https://github.com/rokath/trice/releases/tag/v0.57.0)). Check [triceAuxiolary.c](https://github.com/rokath/trice/blob/master/src/box/triceAuxiolary.c) and `#define TRICE_AUXILIARY` inside [triceConfig.h](https://github.com/rokath/trice/blob/master/test/MDK-ARM_STM32F030R8_instrumented/Core/Inc/triceConfig.h). That part is not tested yet.

**Reply by FabienLep** [reply]
Thanks Thomas!
We did fork and add a custom_write() in TriceWriteDevice(). We added an API to pass a callback to the trice lib and custom_write() calls that callback. This way there isn't any need to change the trice code further if the output needs to be changed.
I can at some point cleanup and propose the changes here if you're interested

**Comment by rokath** [comment]
Hi Fabien, is there anything you propose to add to or change concerning the current taget code to achieve your goals?

**Reply by FabienLep** [reply]
Hi Thomas,
We followed your initial idea and extended TriceWriteDevice with a custom_write() function that calls a callback. We added an API to add that callback. We're pretty happy with that implementation. Do you want me to PR this change back to your project?

**Comment by rokath** [comment]
Sorry, Fabien, I just saw your answer. PRs always welcome! Please check against the latest version first. For example there is now a file `https://github.com/rokath/trice/blob/master/src/triceAuxiliary.c` since a while.
